lib/gui/lvgl: Add user config file

krip-tip opened the pull request:
Add custom preferences file, for example
to enable custom fonts and change preferences.

Signed-off-by: Krivorot Oleg <krivorot.oleg@gmail.com>

stephanosio commented on lib/gui/lvgl/Kconfig:
```diff
@@ -1,5 +1,6 @@
 # Copyright (c) 2018-2019 Jan Van Winkel <jan.van_winkel@dxplore.eu>
 # Copyright (c) 2020 Teslabs Engineering S.L.
+# Copyright (c) 2020 Krivorot Oleg <krivorot.oleg@gmail.com>
```
2021?

oleg-krv replied:
ок

stephanosio commented on lib/gui/lvgl/Kconfig:
```diff
@@ -21,6 +22,14 @@ config APP_LINK_WITH_LVGL
 
+config LVGL_USER_CONFIG_FILE
+    bool "Use user config file"
+    default n
```
All boolean configs are `n` by default. This is not needed.

oleg-krv replied:
ok

stephanosio commented on lib/gui/lvgl/Kconfig:
```diff
@@ -21,6 +22,14 @@ config APP_LINK_WITH_LVGL
 	  disabled if the include paths for LVGL are causing aliasing
 	  issues for 'app'.
 
+config LVGL_USER_CONFIG_FILE
```
Suggestion: `LVGL_USER_CONFIG_HEADER` or `LVGL_USER_CONFIG_INCLUDE` would make it more clear as to what the "config file" is supposed to be.

oleg-krv replied:
it would probably be correct to name  LVGL_USE_USER_CONFIG ?

stephanosio replied:
That does sound more reasonable.

`LVGL_USE_USER_CONFIG`, `LVGL_INCLUDE_USER_CONFIG_HEADER`, etc.

stephanosio commented on lib/gui/lvgl/Kconfig:
```diff
@@ -21,6 +22,14 @@ config APP_LINK_WITH_LVGL
+    bool "Use user config file"
+    default n
+
+config LVGL_USER_CONFIG_FILE_INCLUDE
```
Suggestion: `LVGL_USER_CONFIG_FILE_NAME` (or, if we decide to change the config name above, something equivalent) would make it more clear as to what this config is supposed to describe.

oleg-krv replied:
it would probably be correct to name LVGL_USER_CONFIG_FILE, since the full name of the include file is stored in this variable

stephanosio replied:
If `LVGL_USER_CONFIG_FILE` is renamed to `LVGL_USE_USER_CONFIG`, sure.

stephanosio commented on lib/gui/lvgl/lv_conf.h:
```diff
@@ -871,4 +872,8 @@ typedef void *lv_obj_user_data_t;
 #define LV_USE_WIN IS_ENABLED(CONFIG_LVGL_USE_WIN)
 
+#ifdef CONFIG_LVGL_USER_CONFIG_FILE
+#include CONFIG_LVGL_USER_CONFIG_FILE_INCLUDE
```
Where is the `lv_conf_user.h` supposed to be located?

Is the user application supposed to expose the directory containing one as a global include dir in `CMakeLists.txt`?

oleg-krv replied:
I have it in the root folder of my project, or in any folder, the path to which is registered in CMakeLists.txt via zephyr_include_directories.

stephanosio commented on lib/gui/lvgl/Kconfig:
```diff
@@ -21,6 +22,13 @@ config APP_LINK_WITH_LVGL
 	  disabled if the include paths for LVGL are causing aliasing
 	  issues for 'app'.
 
+config LVGL_USE_USER_CONFIG
+    bool "Use user config file"
+
+config LVGL_USER_CONFIG_FILE
```
These configs should come with `help` describing what they do, how they are supposed to be used, etc.

oleg-krv replied:
```
config LVGL_USE_USER_CONFIG
    bool "Use user config file"
    help
      Add custom preferences file, for example to enable custom fonts.
      In the file, you can override all values set by variables from
      Kconfig or set new ones that cannot be defined using variables
      from Kconfig.

config LVGL_USER_CONFIG_FILE
    string "User config file name"
    default "lv_conf_user.h"
    depends on LVGL_USE_USER_CONFIG
    help
      This file should be located in the root folder of the project,
      or in any folder, the path to which is registered in root
      CMakeLists.txt using the macro zephyr_include_directories()
```
Sorry for my english, are the variable assignments clear enough?

stephanosio reviewed (changes requested):
Please amend commits instead of adding new ones for the changes.

Commit:
lib/gui/lvgl: Add user config file

Add custom preferences file, for example
to enable custom fonts and change preferences.

Signed-off-by: Krivorot Oleg krivorot.oleg@gmail.com

stephanosio commented:
https://docs.zephyrproject.org/latest/contribute/index.html#contribution-workflow

Please see #15.

oleg-krv commented:
> https://docs.zephyrproject.org/latest/contribute/index.html#contribution-workflow
> 
> Please see `#15`.

I'm sorry, I think I figured out how to change and not produce commits. This is my first pull request for a large project

gmarull reviewed (changes requested):
The same behavior can be achieved by setting Kconfig options in `prj.conf`.

oleg-krv commented:
> The same behavior can be achieved by setting Kconfig options in `prj.conf`.

This is if you define the standard options in Kconfig. But there are options that need to be written in lv_config.h so that the lvgl library itself can use them. For example, I have not found any other way to completely replace the standard fonts and use several of my fonts in my theme. All of this is simply solved by including an additional header file at the end.
And in the example from lvgl itself, the user's fonts are included in the configuration file.
it will be more difficult to implement the mechanism for connecting fonts via commands in Kconfig.
And in my program self-written lvgl objects are defined. which are also written to the settings file and collected together with the library.

gmarull commented:
> > The same behavior can be achieved by setting Kconfig options in `prj.conf`.
> 
> This is if you define the standard options in Kconfig. But there are options that need to be written in lv_config.h so that the lvgl library itself can use them. For example, I have not found any other way to completely replace the standard fonts and use several of my fonts in my theme. All of this is simply solved by including an additional header file at the end.
> And in the example from lvgl itself, the user's fonts are included in the configuration file.
> it will be more difficult to implement the mechanism for connecting fonts via commands in Kconfig.
> And in my program self-written lvgl objects are defined. which are also written to the settings file and collected together with the library.

There is no need to add an extra header for custom fonts. The auto-generated C font files can be added to the application (`target_sources(app PRIVATE my_font.c)`) and then you can just use them using `LV_FONT_DECLARE(my_font)`. You can create application level Kconfig options for custom fonts if needed.

oleg-krv commented:
> > > The same behavior can be achieved by setting Kconfig options in `prj.conf`.
> > 
> > 
> > This is if you define the standard options in Kconfig. But there are options that need to be written in lv_config.h so that the lvgl library itself can use them. For example, I have not found any other way to completely replace the standard fonts and use several of my fonts in my theme. All of this is simply solved by including an additional header file at the end.
> > And in the example from lvgl itself, the user's fonts are included in the configuration file.
> > it will be more difficult to implement the mechanism for connecting fonts via commands in Kconfig.
> > And in my program self-written lvgl objects are defined. which are also written to the settings file and collected together with the library.
> 
> There is no need to add an extra header for custom fonts. The auto-generated C font files can be added to the application (`target_sources(app PRIVATE my_font.c)`) and then you can just use them using `LV_FONT_DECLARE(my_font)`. You can create application level Kconfig options for custom fonts if needed.

I agree with the help of the specified definitions and the indication void * lv_theme_default_font_small_custom_ptr and others, I managed to connect my fonts.
But what cannot be done: connect CONFIG_LVGL_USE_THEME_CUSTOM.
Do you have an example?
defined:
CONFIG_LVGL_USE_THEME_CUSTOM = y
CONFIG_LVGL_THEME_CUSTOM_INIT_FUNCTION = "lv_theme_my_init"
but it fails to compile:
lv_conf.h: 280: 1: error: unknown type name 'lv_theme_t'
   280 | lv_theme_t * LVGL_THEME_CUSTOM_INIT_FUNCTION (

but if in lv_conf.h. or in my suggested file ask:
#define LV_THEME_DEFAULT_FLAG 0
#define LV_THEME_DEFAULT_INCLUDE <lv_theme_my.h>
#define LV_THEME_DEFAULT_INIT lv_theme_my_init
everything compiles fine.

There is a proposal to modify the connection of a custom theme
I was unable to pass the name of the initialization function from Kconfig, since all text constants are always quoted there when they are passed to autoconf.h

oleg-krv commented:
and the custom theme through my proposal is connected as a hack, I do not type a custom theme and then redefined the variables :-(

gmarull commented:
@oleg-krv you're right, the Kconfig custom options do not work (Kconfig options expand as a string). An issue should be opened to fix/remove that.

What I've done for custom themes in the past is:

1. create a theme using something like https://github.com/lvgl/lvgl/blob/master/src/extra/themes/basic/lv_theme_basic.c
2. Set the theme when app is started, ie
```
lv_theme_set_act(my_theme_init())
```

oleg-krv commented:
> @oleg-krv you're right, the Kconfig custom options do not work (Kconfig options expand as a string). An issue should be opened to fix/remove that.
> 
> What I've done for custom themes in the past is:
> 
> 1. create a theme using something like https://github.com/lvgl/lvgl/blob/master/src/extra/themes/basic/lv_theme_basic.c
> 2. Set the theme when app is started, ie
> 
> ```
> lv_theme_set_act(my_theme_init())
> ```

This can be done, but how to remove the initialization of the standard theme so that it does not take up memory?

Alternative. Replace block:
```
#if CONFIG_LVGL_USE_THEME_CUSTOM

lv_theme_t *LVGL_THEME_CUSTOM_INIT_FUNCTION(
	lv_color_t color_primary, lv_color_t color_secondary, uint32_t flags,
	const lv_font_t *font_small, const lv_font_t *font_normal,
	const lv_font_t *font_subtitle, const lv_font_t *font_title);

#define LV_THEME_DEFAULT_INIT LVGL_THEME_CUSTOM_INIT_FUNCTION

#endif
```
to this code:
```
#if CONFIG_LVGL_USE_THEME_CUSTOM

#define LV_THEME_DEFAULT_FLAG 0
#define LV_THEME_DEFAULT_INCLUDE  CONFIG_LVGL_CUSTOM_THEME_INCLUDE

#ifndef LV_THEME_DEFAULT_INIT
#error "Define LV_THEME_DEFAULT_INIT in CMake or custom theme header file. In the variable, specify the name of the custom theme initialization function."
#endif

#endif
```
also a hack, but more elegant :-)
